docs/functional_architecture.md

```diff
@@ -0,0 +1,104 @@
+
+## Foreword
+
+At a high level, warc2zim is a piece of software capable to transform a set of WARC files into one ZIM file. From a
+functional point of view, it is hence a "format converter".
+
+While warc2zim is typically used as a sub-component of zimit, where WARC files are produced by Browsertrix crawler, it
+is in fact agnostic of this fact and could process any WARC file adhering to the standard.
+
+This documentation will describe the big functions achieved by warc2zim codebase. It is important to note that these functions
+ are not seggregated inside the codebase with frontiers.
+
+## ZIM storage
+
+While storing the web resources in the ZIM is mostly straightforward (we just transfer the raw bytes, after some
+ modification for URL rewriting if needed), the decision of the path where the resource will be stored is very important.
+
+This is purely conventional, even if ZIM specification has to be respected for proper operation in readers.
+
+This function is responsible to compute the ZIM path where a given web resource is going to be stored.
+
+While the URL is the only driver of this computation for now, warc2zim might have to consider other contextual data in the future.
+E.g. the resource to serve might by dynamic, depending not only on URL query parameters but also header(s) value(s).
+
+## Fuzzy rules
+
+Unfortunately, it is not always possible / desirable to store the resource with a simple transformation.
+
+A typical situation is that some query parameters are dynamically computed by some Javascript code to include user
+ tracking identifier, current datetime information, ...
+
+When running again the same javascript code inside the ZIM, the URL will hence be slightly different because context has changed,
+but the same content needs to be retrieved.
+
+warc2zim hence relies on fuzzy rules to transform/simplify some URLs when computing the ZIM path.
+
+## URL Rewriting
+
+warc2zim transforms (rewrites) URLs found in documents (HTML, CSS, JS, ...) so that they are usable inside the ZIM.
+
+### General case
+
+One simple example is that we might have following code in an HTML document to load an image with an absolute URL:
+
+```
+  <img src="https://en.wikipedia.org/wiki/File:Kiwix_logo_v3.svg"></img>
+```
+
+The URL `https://en.wikipedia.org/wiki/File:Kiwix_logo_v3.svg` has to be transformed to a URL that it is usable
+inside the ZIM.
+
+Since the ZIM will typically be served by kiwix-serve, this has to be a relative URL (other readers follow the same kind of conventions
+on how to access resources inside a ZIM from a given web resource). This relative URL is hence dependant on the
+location of the resource currently being rewriten.
+
+The table below gives some examples of what the rewritten URL is going to be, depending on the URL of the rewritten document.
+
+|--|--|
+| HTML document URL | image URL rewritten for usage inside the ZIM |
+| `https://en.wikipedia.org/wiki/Kiwix` | `./File:Kiwix_logo_v3.svg` |
```

Comment:
Pretty sure examples are incorrect and thus misleading

Reply:
I'm pretty sure it is correct ^^

Could you point me to what seems incorrect? (or should we discuss it live?)

Reply:
Nevermind ; toggling between mardown source and preview confused me 😅